pkg/erofs/erofs.go

```diff
@@ -690,87 +724,155 @@ func (i *Inode) blocks() uint64 {
+	for left <= right {
+		mid := uint64(left+right) >> 1
+		direntOff := target.block.base + mid*DirentSize
 		d, err := i.image.direntAt(direntOff)
 		if err != nil {
-			return err
+			return 0, err
 		}
-		// Apart from the offset of the first filename, nameOff0 also indicates
-		// the total number of dirents in this block.
-		nameOff0 := start + uint64(d.NameOff)
-		maxSize := blockSize
-		if blocks == 1 {
-			if tailSize := uint32(i.size) & (blockSize - 1); tailSize != 0 {
-				maxSize = tailSize
-			}
+		name, err := i.getDirentName(d, target.block, uint16(mid) == target.numDirents-1)
+		if err != nil {
+			return 0, err
 		}
-		// Iterate all the dirents in this block.
-		for d != nil {
-			var (
-				next    *Dirent
-				nameLen uint32
-			)
-			direntOff += uint64(d.SizeBytes())
-			lastDirent := direntOff >= nameOff0
-			if lastDirent {
-				// There is no more dirent in this block, d is the last one.
-				next = nil
-				nameLen = maxSize - uint32(d.NameOff)
-			} else {
-				// Get the next adjacent dirent.
-				if next, err = i.image.direntAt(direntOff); err != nil {
-					return err
-				}
-				nameLen = uint32(next.NameOff - d.NameOff)
-			}
+		switch bytes.Compare(nameBytes, name) {
+		case 0:
+			// Found the target dirent.
+			return d.Nid(), nil
+		case 1:
+			left = int64(mid) + 1
+		case -1:
+			right = int64(mid) - 1
+		}
+	}
+
+	return 0, linuxerr.ENOENT
+}
 
-			buf, err := i.image.BytesAt(start+uint64(d.NameOff), uint64(nameLen))
+// IterDirents invokes cb on each entry in the directory represented by this inode.
+// The directory entries will be iterated in alphabetical order.
+func (i *Inode) IterDirents(cb func(name string, typ uint8, nid uint64) error) error {
+	if !i.IsDir() {
+		return linuxerr.ENOTDIR
+	}
+
+	// Iterate all the blocks which contain dirents.
+	for blockIdx := uint64(0); blockIdx < i.blocks; blockIdx++ {
+		block := i.getBlockDataInfo(blockIdx)
+		d, err := i.getDirent0(block)
+		if err != nil {
+			return err
+		}
+		// Iterate all the dirents in this block.
+		for numDirents := d.NameOff / DirentSize; numDirents > 0; numDirents-- {
+			name, err := i.getDirentName(d, block, numDirents == 1)
 			if err != nil {
 				return err
 			}
-			if lastDirent {
-				if n := bytes.IndexByte(buf, 0); n != -1 {
-					nameLen = uint32(n)
-				}
-			}
-			if nameLen > MaxNameLen {
-				log.Warningf("Corrupted dirent at inode (nid=%v)", i.Nid())
-				return linuxerr.EUCLEAN
-			}
-			name := string(buf[:nameLen])
-			if err := cb(name, d.FileType, d.Nid()); err != nil {
+			if err := cb(string(name), d.FileType, d.Nid()); err != nil {
 				return err
 			}
-
-			// Go on to process the next adjacent dirent.
-			d = next
-		}
-
-		blocks--
-
-		if blocks == 1 && i.idataOff != 0 {
-			// If we have any inline data and this is the last block, we need to process it now.
-			start = i.idataOff
-		} else {
-			// Just go on to process the next adjacent block.
-			start += uint64(blockSize)
+			d = direntAfter(d)
```

Comment:
During last iteration, this breaks the precondition of `direntAfter()` that `d` is not the last dirent.

Reply:
Good catch. Will fix. Thanks!